### ISSUE TITLE
Loaders merge behavior with "include"

### ISSUE THREAD

**Issue by VladimirGrenaderov** [survivejs/webpack-merge#68]
I found issue with merging multiple loaders (in 2.6.1) with same test and different include fields:

rule 1: { test: 'js', <global loaders>} // <-- no include field
....
rule 10: { test: 'js', include: 'verySpecialDir', 'loaders only for this dir'}
rule 11: { test: 'js', include: 'anotherSpecialDir', 'different loaders'}

merge.smart will return:
{ test: 'js', include: 'verySpecialDir', 'loaders only for this dir' merged with 'global loaders'}
{ test: 'js', include: 'anotherSpecialDir', 'different loaders'}

At first, rule 1 merged with rule 10, because of rule 10 looks like it extend rule 1 with include field. Next, rule 10 not merged with rule 11 - they had different include fields. As result, global rule applied only for files in one dir, not for all js.

Yes, it's expected behavior (one rule extends another), but it's looks not really 'smart'. Currently, I use slightly different (and useless for webpack) test fields to avoid merge, but it's very bad idea, I think.

Write a customizeArray is slightly better option, but it's very tricky and unclear solution at my sight.

May be smart.merge should not merge include fields (exclude, and other scope modificators) by default, and treat loaders with different includes as different objects?

**Comment by bebraw** [comment]
I think I'll deprecate `merge.smart` in 3.0. Too unpredictable.

**Comment by VladimirGrenaderov** [comment]
But how to replace current smart behavior? It's need to write entry with merge.unique and compare test, include, etc in every project?

**Comment by bebraw** [comment]
I guess the quickest fix would be to **not** merge when there's no `include`. That would be the safest option.

**Comment by bebraw** [comment]
Any chance for a test case?

```javascript

it('should not merge to parent without include ' + loadersKey, function () {
  const common = {
    module: {}
  };
  common.module[loadersKey] = [
    {
      test: /\.js$/,
      loaders: ['babel']
    }
  ];
  const strip = {
    module: {}
  };
  strip.module[loadersKey] = [
    {
      test: /\.js$/,
      loaders: ['strip?strip[]=debug'],
      include: 'apps'
    }
  ];
  const result = {
    module: {}
  };
  result.module[loadersKey] = [
    {
      test: /\.js$/,
      loaders: ['babel']
    },
    {
      test: /\.js$/,
      loaders: ['strip?strip[]=debug'],
      include: 'apps'
    }
  ];

  assert.deepEqual(merge(common, strip), result);
});
```

passes for me even though based on your description it should not. I'm missing something in my test.

**Comment by VladimirGrenaderov** [comment]
My bad - order in example does matter. Fixed example:
```javascript
const merge = require('webpack-merge');

var configs = [
    {
        module: {
            rules: [
                {
                    test: /\.js$/,
                    include: 'dir-1',
                    use: [
                        'custom-loader-1'
                    ]
                }
            ]
        }
    }, {
        module: {
            rules: [
                {
                    test: /\.js$/,
                    use: [
                        'common-loader'
                    ]
                }
            ]
        }
    }, {
        module: {
            rules: [
                {
                    test: /\.js$/,
                    include: 'dir-2',
                    use: [
                        'custom-loader-2'
                    ]
                }
            ]
        }
    }
];

console.log(JSON.stringify(merge.smart(configs), null, '\t'));
```

Result:
```json
{
	"module": {
		"rules": [
			{
				"test": {},
				"include": "dir-1",
				"use": [
					"custom-loader-1",
					"common-loader"
				]
			},
			{
				"test": {},
				"include": "dir-2",
				"use": [
					"custom-loader-2"
				]
			}
		]
	}
}
```

**Comment by bebraw** [comment]
That helps. Thanks. 👍

**Comment by bebraw** [comment]
Hi,

I have an initial fix at https://github.com/survivejs/webpack-merge/pull/69 if you want to have a look.

I'll likely merge that in a few days and publish a version with the fix.

**Comment by VladimirGrenaderov** [comment]
Looks good, thanks!

**Comment by bebraw** [comment]
Published in `3.0.0`. Sorry for the delay, busy days.

**Comment by VladimirGrenaderov** [comment]
Just upgraded to 3.0.0 - works perfectly, thank!
